### PR TITLE
Thread names in logs and deadlock debug tools

### PR DESCRIPTION
**Pull request by jamesob**
I'm resurrecting this one (from #13168) because I need it to make progress on #15735. 

It's now off by default and can be turned on with `-logthreadnames=1`.

Ran some benchmarks (IBD from local peer from 500_000 -> 504_000) and it's within spitting distance either on or off:

#### 2018-05-threadnames.3 vs. master (absolute)
|                      name                      | iterations |   2018-05-threadnames.3    |           master           |
|------------------------------------------------|-----------:|----------------------------|----------------------------|
| ibd.local.500000.504000.dbcache=2048           |          3 | 376.1584 (± 9.2944)        | 392.3414 (± 13.4238)       |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 | 2236117.3333 (± 1845.9623) | 2238690.6667 (± 2669.3487) |

#### 2018-05-threadnames.3 vs. master (relative)
|                      name                      | iterations | 2018-05-threadnames.3 | master |
|------------------------------------------------|-----------:|----------------------:|-------:|
| ibd.local.500000.504000.dbcache=2048           |          3 |                     1 |  1.043 |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 |                     1 |  1.001 |

#### 2018-05-threadnames-take-2 vs. master (absolute)
|                      name                      | iterations | 2018-05-threadnames-take-2 |           master           |
|------------------------------------------------|-----------:|----------------------------|----------------------------|
| ibd.local.500000.504000.dbcache=2048           |          3 | 367.6861 (± 0.3941)        | 364.1667 (± 0.9776)        |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 | 2238461.3333 (± 3697.8730) | 2237014.6667 (± 3307.6966) |

#### 2018-05-threadnames-take-2 vs. master (relative)
|                      name                      | iterations | 2018-05-threadnames-take-2 | master |
|------------------------------------------------|-----------:|---------------------------:|-------:|
| ibd.local.500000.504000.dbcache=2048           |          3 |                      1.010 |   1.00 |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 |                      1.001 |   1.00 |
```

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

File: src/logging.cpp [hunk 1]
```diff
@@ -196,29 +199,31 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
...
     return strStamped;
 }
 
 void BCLog::Logger::LogPrintStr(const std::string &str)
 {
-    std::string strTimestamped = LogTimestampStr(str);
+    std::string strPrefixed = str;
+
+    if (m_log_threadnames && m_started_new_line)
+        strPrefixed.insert(0, "[" + std::move(thread_util::GetInternalName()) + "] ");
```
Comment (practicalswift) [inline src/logging.cpp]: I think this is a pessimizing move which prevents copy elision.

File: src/sync.cpp [hunk 2]
```diff
@@ -37,23 +39,23 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
...
-    CLockLocation(const char* pszName, const char* pszFile, int nLine, bool fTryIn)
-    {
-        mutexName = pszName;
-        sourceFile = pszFile;
-        sourceLine = nLine;
-        fTry = fTryIn;
-    }
+    CLockLocation(
+        const char* pszName, const char* pszFile, int nLine, bool fTryIn, std::string thread_name_) :
+            mutexName(pszName), sourceFile(pszFile), sourceLine(nLine), fTry(fTryIn),
```
Comment (practicalswift) [inline src/sync.cpp]: Reorder to match the order the compiler will use (which is the order of declaration): so please put `fTry` first :-)

File: src/logging.cpp (hunk 1 above)
Comment (jamesob) [reply src/logging.cpp]: Fixed, thanks.

File: src/sync.cpp (hunk 2 above)
Comment (jamesob) [reply src/sync.cpp]: Fixed, thanks.

File: src/sync.cpp [hunk 3]
```diff
@@ -37,23 +39,26 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
...
+        return tfm::format(
+            "%s %s:%s%s (in thread %s)",
+            mutexName, sourceFile, itostr(sourceLine), (fTry ? " (TRY)" : ""), thread_name);
     }
 
 private:
     bool fTry;
     std::string mutexName;
     std::string sourceFile;
+    std::string thread_name;
```
Comment (ryanofsky) [inline src/sync.cpp]: In commit "threads: add thread names to deadlock debugging message" (503c1ff6eeeb6e222ff446033745650ba59ab659)

In case you do take my earlier suggestion and change `GetInternalName` to return a const string reference instead of a temporary string, you could change this type to a const string reference and avoid any string copy initializing this struct field (it would point directly to the `g_thread_name` instance).

File: src/logging.cpp [hunk 4]
```diff
@@ -174,7 +177,7 @@ std::vector<CLogCategoryActive> ListActiveLogCategories()
     return ret;
 }
 
-std::string BCLog::Logger::LogTimestampStr(const std::string &str)
+std::string BCLog::Logger::LogTimestampStr(const std::string str)
```
Comment (ryanofsky) [inline src/logging.cpp]: In commit "threads: prefix log messages with thread names" (4fba3ead9811ebe47161e9e7f535a7cc88403cf5)

Not sure if you dropped the `&` intentionally here. It would make sense to drop it, so the string could be moved from instead of copied. But you'd also want to drop the `const` and add `std::move` below, and also change the `Logger::LogPrintStr` function similarly (function calling this function).

Without these things, I don't think there'd be a reason to drop `&`.

File: src/logging.cpp [hunk 5]
```diff
@@ -196,29 +199,31 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
...
 {
-    std::string strTimestamped = LogTimestampStr(str);
+    std::string strPrefixed = str;
+
+    if (m_log_threadnames && m_started_new_line)
+        strPrefixed.insert(0, "[" + thread_util::GetInternalName() + "] ");
+
+    strPrefixed = LogTimestampStr(strPrefixed);
+
+    m_started_new_line = (!str.empty() && str[str.size()-1] == '\n');
```
Comment (ryanofsky) [inline src/logging.cpp]: In commit "threads: prefix log messages with thread names" (4fba3ead9811ebe47161e9e7f535a7cc88403cf5)

Some unneeded parentheses here

File: src/logging.cpp [hunk 6]
```diff
@@ -196,29 +199,31 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
...
-    else
-        m_started_new_line = false;
-
     return strStamped;
 }
 
 void BCLog::Logger::LogPrintStr(const std::string &str)
 {
-    std::string strTimestamped = LogTimestampStr(str);
+    std::string strPrefixed = str;
```
Comment (ryanofsky) [inline src/logging.cpp]: In commit "threads: prefix log messages with thread names" (4fba3ead9811ebe47161e9e7f535a7cc88403cf5)

Since this is renaming, could go to with current lowercase convention.

**Review by ryanofsky — APPROVED** [review]
utACK 503c1ff6eeeb6e222ff446033745650ba59ab659. Looks like some windows tests are failing, but nothing to do with this PR. I left a few suggestions for changes, but please ignore any that do not sound like super great and wonderful ideas.

File: src/sync.cpp (hunk 3 above)
Comment (jamesob) [reply src/sync.cpp]: In making this change I hit some compiler barf that has me out of my depths: https://gist.github.com/jamesob/5a19c75b34a49a4972b4105e0287d79d

**Comment by jamesob** [comment]
@ryanofsky thanks very much for the review. Per usual, your suggested changes are indelibly righteous. I've made all of them save for the last I commented on.

File: src/sync.cpp (hunk 3 above)
Comment (ryanofsky) [reply src/sync.cpp]: Oh, the relevant part of that error message is:

```
sync.cpp:133:33: note: in instantiation of member function 'std::vector<std::pair<void *, CLockLocation>, std::allocator<std::pair<void *, CLockLocation> > >::operator=' requested here
        lockdata.lockorders[p1] = g_lockstack;
```

It's complaining that it can't copy the g_lockstack variable, which makes sense since adding a reference member to CLockLocation makes it not copyable by default. If inclined, you could work around it by manually defining a CLockLocation copy constructor:

```
CLockLocation(const CLockLocation& other) : fTry(other.fTry), mutexName(other.mutexName), etc, ... {}
```

**Comment by jamesob** [comment]
Reran benchmarking; no performance difference.

### absolute
|                      name                      | iterations |           master           |      threadnames2.18       |
|------------------------------------------------|-----------:|----------------------------|----------------------------|
| build.make.23.clang                            |          1 | 135.8741 (± 0.0000)        | 133.9639 (± 0.0000)        |
| build.make.23.clang.mem-usage                  |          1 | 648692.0000 (± 0.0000)     | 646632.0000 (± 0.0000)     |
| ibd.local.500000.504000.dbcache=2048           |          3 | 357.4552 (± 1.2016)        | 357.3936 (± 1.1932)        |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 | 2240597.3333 (± 3291.6968) | 2237538.6667 (± 4462.0143) |

### relative
|                      name                      | iterations | master | threadnames2.18 |
|------------------------------------------------|-----------:|-------:|----------------:|
| build.make.23.clang                            |          1 |  1.014 |               1 |
| build.make.23.clang.mem-usage                  |          1 |  1.003 |               1 |
| ibd.local.500000.504000.dbcache=2048           |          3 |  1.000 |               1 |
| ibd.local.500000.504000.dbcache=2048.mem-usage |          3 |  1.001 |               1 |

File: src/sync.cpp (hunk 3 above)
Comment (ryanofsky) [reply src/sync.cpp]: I think I was wrong. Even with a reference member CLockLocation should still be copyable, just not assignable, so you may be able to fix this by emplacing into lockorders instead of assigning into it.

Comment (jamesob) [reply src/sync.cpp]: Fixed by changing the above assignment to `emplace()` per your advice. Thanks!

File: src/util/threadnames.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+static thread_local std::string g_thread_name;
+const std::string& util::ThreadGetInternalName() { return g_thread_name; }
+//! Set the in-memory internal name for this thread. Does not affect the process
+//! name.
+static void SetInternalName(const std::string& name) { g_thread_name = name; }
+
+// Without thread_local available, don't handle internal name at all.
+#else
+
+const std::string& util::ThreadGetInternalName() { return ""; }
```
Comment (ryanofsky) [inline src/util/threadnames.cpp]: In commit "threads: introduce threadutil, refactor thread naming" (1cf8779ee2887e3c8ea0265fe8bbb9ecbd502172)

Sorry, this is from a change I suggested, but it isn't really safe (and causes a warning): 

```
util/threadnames.cpp:48:59: warning: returning reference to local temporary object [-Wreturn-stack-address]
const std::string& util::ThreadGetInternalName() { return ""; }
```

Should do something like:

```c++
static const std::string empty_string;
static void SetInternalName(const std::string& name) { return empty_string; }
```

File: src/util/threadnames.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_THREADUTIL_H
+#define BITCOIN_THREADUTIL_H
```
Comment (ryanofsky) [inline src/util/threadnames.h]: In commit "threads: introduce threadutil, refactor thread naming" (1cf8779ee2887e3c8ea0265fe8bbb9ecbd502172)

Should use `BITCOIN_UTIL_THREADNAMES_H`. Also commit message text reference previous names and could be updated.

**Review by ryanofsky — APPROVED** [review]
utACK 2e53d900ba40623b203bf21fc572a0fe56669fc7. Just various suggested changes since last review.

**Comment by jamesob** [comment]
@ryanofsky Oops! Thanks, pushed.

**Review by ryanofsky — APPROVED** [review]
utACK eed4a5c7f81b53e47df3c4505f2b273a25fc9980. Reference to temporary bug's fixed now. There is still the lint problem causing travis to fail, though (https://travis-ci.org/bitcoin/bitcoin/jobs/525122780#L356).

File: src/test/util_threadnames_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(util_threadnames_tests, BasicTestingSetup)
+
+std::string TEST_THREAD_NAME_BASE = "test_thread.";
```
Comment (MarcoFalke) [inline src/test/util_threadnames_tests.cpp]: in commit 195ba388eb tests: add threadutil tests


style-nit: Should be `const`, since it is upper cased?

File: doc/release-notes-pr13168.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+Thread names in logs
+--------------------
+
+On platforms supporting `thread_local`, log lines are now prefixed by default
+with the name of the thread that caused the log. To disable this behavior,
+specify `-logthreadnames=0`.
```
Comment (MarcoFalke) [inline doc/release-notes-pr13168.md]: Why does the OP say it is off by default, but the code and release notes otherwise?

```
 static const bool DEFAULT_LOGTHREADNAMES = true;

File: src/logging.cpp [hunk 11]
```diff
@@ -196,29 +199,31 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
...
     return strStamped;
 }
 
 void BCLog::Logger::LogPrintStr(const std::string &str)
 {
-    std::string strTimestamped = LogTimestampStr(str);
+    std::string str_prefixed = str;
+
+    if (m_log_threadnames && m_started_new_line)
+        str_prefixed.insert(0, "[" + util::ThreadGetInternalName() + "] ");
```
Comment (MarcoFalke) [inline src/logging.cpp]: style-nit: in commit 951769cf12 threads: prefix log messages with thread names

Could add `{}`, according to developer notes for new code?

**Review by maflcko — APPROVED** [review]
utACK 043a387639

Just some style comments that should probably be ignored

File: doc/release-notes-pr13168.md (hunk 10 above)
Comment (jamesob) [reply doc/release-notes-pr13168.md]: Wowow, something must've gotten screwed up in a rebase. Fixing - thanks for the catch.

File: src/logging.cpp (hunk 11 above)
Comment (jamesob) [reply src/logging.cpp]: Done, thanks.

File: src/test/util_threadnames_tests.cpp (hunk 9 above)
Comment (jamesob) [reply src/test/util_threadnames_tests.cpp]: Done, thanks.

**Comment by jamesob** [comment]
Latest benchmark; 0.7% difference seems attributable to noise.

ibd local 500000 504000 dbcache=2048

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15849/commits/0cfcb8d470bd23f34e1bd81810444a1db8e4c702

nit: if I'm not mistaken you can use rvalues and move to avoid copies of the name throughout (channeling @sipa) https://github.com/bitcoin/bitcoin/pull/13168#discussion_r186303737

```diff
diff --git a/src/util/threadnames.cpp b/src/util/threadnames.cpp
index 3217855f7..2ca91613c 100644
--- a/src/util/threadnames.cpp
+++ b/src/util/threadnames.cpp
@@ -40,18 +40,18 @@ static thread_local std::string g_thread_name;
 const std::string& util::ThreadGetInternalName() { return g_thread_name; }
 //! Set the in-memory internal name for this thread. Does not affect the process
 //! name.
-static void SetInternalName(const std::string& name) { g_thread_name = name; }
+static void SetInternalName(std::string&& name) { g_thread_name = name; }
 
 // Without thread_local available, don't handle internal name at all.
 #else
 
 static const std::string empty_string;
 const std::string& util::ThreadGetInternalName() { return empty_string; }
-static void SetInternalName(const std::string& name) { }
+static void SetInternalName(std::string&& name) { }
 #endif
 
-void util::ThreadRename(const std::string& name)
+void util::ThreadRename(std::string&& name)
 {
     SetThreadName(("bitcoin-" + name).c_str());
-    SetInternalName(name);
+    SetInternalName(std::move(name));
 }
diff --git a/src/util/threadnames.h b/src/util/threadnames.h
index ecf6c50cc..aaf07b9bf 100644
--- a/src/util/threadnames.h
+++ b/src/util/threadnames.h
@@ -10,7 +10,7 @@
 namespace util {
 //! Rename a thread both in terms of an internal (in-memory) name as well
 //! as its system thread name.
-void ThreadRename(const std::string&);
+void ThreadRename(std::string&&);
 
 //! Get the thread's internal (in-memory) name; used e.g. for identification in
 //! logging.
```

**Comment by jamesob** [comment]
@Empact thanks, pushed.

**Comment by practicalswift** [comment]
I'm not sure the change from `const std::string&` to `std::string&&` is an improvement TBH.

It brings no performance improvement in this case AFAIK and `const std::string&` (851 instances) is what we're using in the rest of the code base.

```
$ git grep -E 'const std::string *&[^&]' | wc -l
851
$ git grep -E 'const std::string *&&' | wc -l
0
```

**Comment by Empact** [comment]
@practicalswift Rvalues make sense in cases where you have setters or constructors which can consume / incorporate the object passed in. We're not using it with string but you can see other uses here:
```
$ git grep -E '[a-z]&&' | wc -l
44
```

If you're not seeing a performance benefit, it could be because my patch is apparently missing one move, as per: https://www.chromium.org/rvalue-references#TOC-9.-Move-constructors-should-be-accompanied-by-move-assignment-operators.
```patch
-static void SetInternalName(std::string&& name) { g_thread_name = name; }
+static void SetInternalName(std::string&& name) { g_thread_name = std::move(name); }
```

I don't think this is critical, but it seems semantically appropriate.

**Review by ryanofsky — APPROVED** [review]
utACK 4b4bc0067b41efaaff01abdaa8974aa7b88c132c. Only changes since last review are disabling by default and adding `const` and `&&`.

---

re: https://github.com/bitcoin/bitcoin/pull/15849#issuecomment-487301553

> I'm not sure the change from const std::string& to std::string&& is an improvement TBH.

I also wouldn't consider this an improvement because now you can't call `SetInternalName` with an lvalue (plain variable or reference). This is now an error:

```
std::string name = "something";
SetInternalName(name);
```

People get really confused about this stuff, but if the goal is to just move-enable `SetInternalName`, it should take plain `std::string` not `const std::string&`, and not `std::string&&`.

- If a function just needs to read an argument (and not manipulate it before throwing it away or move it into a data structure) it should take `const T&`.

- If a function wants to give callers the option of moving from an argument but still allow copies, it should take plain `T`.

- If a function wants to force callers to move it should take `T &&`.

File: src/sync.cpp [hunk 12]
```diff
@@ -37,23 +39,26 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
...
 struct CLockLocation {
-    CLockLocation(const char* pszName, const char* pszFile, int nLine, bool fTryIn)
-    {
-        mutexName = pszName;
-        sourceFile = pszFile;
-        sourceLine = nLine;
-        fTry = fTryIn;
-    }
+    CLockLocation(
+        const char* pszName, const char* pszFile, int nLine, bool fTryIn, const std::string& thread_name) :
```
Comment (MarcoFalke) [inline src/sync.cpp]: style-nit (since you'd have to fix up other things anyway): Could move the colon to the next line, according to the style-guideline for new code?

```suggestion
        const char* pszName, const char* pszFile, int nLine, bool fTryIn, const std::string& thread_name)
            : fTry...
```

File: src/util/threadnames.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+// If we have thread_local, just keep thread ID and name in a thread_local
+// global.
+#if defined(HAVE_THREAD_LOCAL)
+
+static thread_local std::string g_thread_name;
+const std::string& util::ThreadGetInternalName() { return g_thread_name; }
+//! Set the in-memory internal name for this thread. Does not affect the process
+//! name.
+static void SetInternalName(std::string&& name) { g_thread_name = name; }
```
Comment (MarcoFalke) [inline src/util/threadnames.cpp]: I'd prefer to undo this change. This does not give any performance gain, nor is it more readable

File: doc/release-notes-pr13168.md [hunk 14]
```diff
@@ -0,0 +1,6 @@
+Thread names in logs
+--------------------
+
+On platforms supporting `thread_local`, log lines are now prefixed by default
```
[the remark this answers is not in the document]
Comment (jamesob) [reply doc/release-notes-pr13168.md]: Argh, another artifact of a bad rebase. Thanks.

File: src/util/threadnames.cpp (hunk 13 above)
Comment (jamesob) [reply src/util/threadnames.cpp]: Yep, reverting.

Comment (jamesob) [reply src/util/threadnames.cpp]: FWIW all this bikeshedding is happening over a function call that is made ~20 times throughout the entire run of `bitcoind`.

**Comment by jamesob** [comment]
Thanks for the input, all. I've reverted back to Russ' case 2 with a `std::move` addition. In any case it doesn't matter much since this function is only called a handful of times during a bitcoind run.

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15849/commits/8722e54e56fd959fd4ff2321b36a7640dee440c5

**Comment by jnewbery** [comment]
This is great. Very nicely structured PR and excellent commit messages 🙌 

utACK 8722e54e56fd959fd4ff2321b36a7640dee440c5

The functional tests should be updated to have this enabled by default: https://github.com/bitcoin/bitcoin/blob/10ed4dff24470f22b4bc4030d534f4e1ed5f1c2f/test/functional/test_framework/test_node.py#L83

**Comment by MarcoFalke** [comment]
re-utACK 8722e54e56fd959fd4ff2321b36a7640dee440c5 (Only change since my previous review is DEFAULT_LOGTHREADNAMES=false and stylistic updates

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-utACK 8722e54e56fd959fd4ff2321b36a7640dee440c5 (Only change since my previous review is DEFAULT_LOGTHREADNAMES=false and stylistic updates
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgtewv9FxicQmO1+eKzOzgPMHm+KzOsWSRPTJVEtCtLgpSOpE2p8YAYCtxnQQEi
XI+7EzSSyEE9IL66WPwXiDVxG3RkZcCpgKcf/oIb5CaVxRN9Ob03uYNup9qCOAey
YZgg9GMkhbAxfOh5ZO3LbembFqMYBv1KqHnSuGSil2VAnaJdTFY6IMhhwR5SWvZs
1Peh510KITCYCg1doua9FcPUBB67yG3zqF36PlHUGivA5C+Zr9kLMNSdNbBzSXf5
xlsqimIGSRjuk9qQx5P31XLx+yPEn3Si1ruTiN85CYMiPdRgIzg9lOmGHSBh1g38
gV19YXiUjikxSWBndXKht9ensL6SPnxseD5Zsca1qAaKbcG+V6rkgE8sLEZro4Qn
QShClrPnlwXvVmdJ8Ly5bnkyYWtC4jhlHu0bdtiyl/NOOWr29x6hGDsWw5l8tGfX
anwN/XWxYIn4pC5La3LP/biD4Jd/vrp0kL1ENp3723OxvP/iiHh+6E/6EcD4stBE
vv147CdK
=4CZG
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `191793320e418666d66ac3696420980055c8e7f889470d345e05c3e4669b7398  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108191793320e418666d66ac3696420980055c8e7f889470d345e05c3e4669b7398f010265ea4a3c120652c4781aab60ca2ec5d08fff0103e0651e85e394d103c57b3dc3b70b86c08f01040a28e9f8766cde526fc52b3772a633708f1045cc8a0d2f00830368206409c34940083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0108e01e9950408ba7460157cefbc480b7c08f1045cc8a0d0f0083b4d236e03ae76df0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d5cd9a753b63b0dc21afd8ed11e6073108f1045cc8a0d1f0081886bfce3b98b5830083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: configure.ac [hunk 15]
```diff
@@ -827,8 +827,23 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
...
+     *mingw*)
+        # mingw32's implementation of thread_local has also been shown to behave
+        # erroneously under concurrent usage; see:
+        # https://gist.github.com/jamesob/fe9a872051a88b2025b1aa37bfa98605
+        AC_MSG_RESULT(no)
+        ;;
+      *darwin*)
+        # TODO enable thread_local on later versions of Darwin where it is
+        # supported (per https://stackoverflow.com/a/29929949)
+        AC_MSG_RESULT(no)
```
Comment (jonasschnelli) [inline configure.ac]: Why can't we check with something like:

```bash
AC_MSG_CHECKING([for thread_local support])
AC_LINK_IFELSE([AC_LANG_SOURCE([
  #include <thread>
  static thread_local int foo = 0;
  static void run_thread() { foo++;}
  int main(){
  for(int i = 0; i < 10; i++) { std::thread(run_thread).detach();}
  return foo;
  }
  ])],
  [
    AC_DEFINE(HAVE_THREAD_LOCAL,1,[Define if thread_local is supported.])
    AC_MSG_RESULT(yes)
  ],
  [
    AC_MSG_RESULT(no)
  ]
)
```

Comment (jamesob) [reply configure.ac]: Because bad `thread_local` behavior sometimes doesn't fail explicitly on certain platforms (e.g. mingw32) until execution, and only under certain circumstances.

Comment (jonasschnelli) [reply configure.ac]: Can you elaborate a little bit what "bad behavior" exactly mean?
I assume the AC_LINK_IFELSE test would work on OSX,.. right? Or would it be a problem for our depends OSX build where we build with SDK 10.11 but allow min os version of 10.10?

Maybe I just open a PR and test it a bit.

Comment (MarcoFalke) [reply configure.ac]: Yeah, I couldn't add thread_local to the tests because of issues with macos: https://github.com/bitcoin/bitcoin/pull/14985

Might be solved by your AC_LINK check, but disabling it seems safer than running into issues later on.

Comment (jamesob) [reply configure.ac]: @jonasschnelli see the gist I link to in the code comment above (https://gist.github.com/jamesob/fe9a872051a88b2025b1aa37bfa98605). This example compiles but stackoverflows when run, and I had to specially construct that example IIRC because trivial usages worked okay.
